### PR TITLE
add a way to always force option notifications

### PR DESCRIPTION
**Pull request by Dudemanguy**
Currently in mpv, option notifications are only propagated if the value changes, but there are times where we always want the notification regardless of value. This adds a simple flag to the `m_option` that forces the option to send an update notification on any write. A couple of options that can benefit from this are `window-scale` and `geometry` so add them here.

### CODE REVIEW

**Comment by Dudemanguy** [comment]
Thanks. I totally forgot about the paragraphs I wrote in `input.rst`.

**Comment by guidocella** [comment]
This is also useful for `--geometry`.

**Comment by Dudemanguy** [comment]
Good point. `geometry`/`autofit` in general would benefit.

**Comment by Dudemanguy** [comment]
>This is also useful for --geometry.

I implemented this, but it ended up being surprisingly convoluted in order to get the `m_config_cache_update` to work. Everything in that file is pretty complicated though.

Also writing the same value twice in a row doesn't work on x11 with this because of the way the runtime geometry updates are implemented. I'll look into it later but the option notification is triggered.

**Comment by Dudemanguy** [comment]
>Also writing the same value twice in a row doesn't work on x11 with this because of the way the runtime geometry updates are implemented.

Should work now. I went with the simpler change instead of trying to refactor more stuff.

**Comment by kasper93** [comment]
Is forcing options really intended? I think it makes sense to do nothing, if unchanged. I understand reasoning, but feels bit off.

Maybe for those where it is needed we should keep `current` property and instead force always, check this property.

Meaning, keep having `current-window-scale` property and `window-scale` option and tie them together in runtime.
`{"window-scale", OPT_DOUBLE(window_scale), M_RANGE(0.001, 100), .update_if_different = "current-window-scale"},`
or something.

**Comment by Dudemanguy** [comment]
>Maybe for those where it is needed we should keep current property and check this property first?

A lot easier said than done. You would need to design some way for the core option code to "reach" into the property execution stuff and be able to decide if it was different from the previous state and update it. The option code doesn't have access to `MPContext` which complicates things even more.

**Comment by kasper93** [comment]
> A lot easier said than done. You would need to design some way for the core option code to "reach" into the property execution stuff

Fair, I don't know the details here, maybe it is tricky.

>  and be able to decide if it was different from the previous state and update it.

I mean, it would compare `window-scale` option with current `current-window-scale` property. So no need to know if property itself changed, but if it is different from option that is being set, even if the option itself has not changed since last set.

But ok, I agree if it is not practical to do we don't have to do it.

**Comment by Dudemanguy** [comment]
>I mean, it would compare window-scale option with current current-window-scale property.

Ah true for window scale it wouldn't be so complicated. I was thinking of `geometry`.

File: options/m_config_core.c [hunk 1]
```diff
@@ -635,12 +658,26 @@ static void update_next_option(struct m_config_cache *cache, void **p_opt)
                         *p_opt = ddst;
                         return;
                     }
+
+                    if (opt->force_update && gsrc->force_updates_len) {
```
[the remark this answers is not in the document]
Comment (Dudemanguy) [reply options/m_config_core.c:662]: I think I avoided it because I thought it was ugly but yeah I can do that.

**Comment by Dudemanguy** [comment]
I undid the writing to `current-window-scale` deprecation in this since it's evidently still useful in some cases.

File: DOCS/interface-changes/deprecate-current-window-scale-write.txt [hunk 2]
```diff
@@ -0,0 +1 @@
+deprecate writing to `current-window-scale` (use `window-scale` instead)
```
Comment (kasper93) [inline DOCS/interface-changes/deprecate-current-window-scale-write.txt:1]: Is this still valid?

File: DOCS/man/input.rst [hunk 3]
```diff
@@ -2734,10 +2710,9 @@ Property list
     calculated from the last non-fullscreen size of the window. The property
     is unavailable if no video is active.
 
-    When setting this property in the fullscreen or maximized state, the behavior
-    is the same as window-scale. In all other cases, setting the value of this
-    property will always resize the window. This does not affect the value of
-    ``window-scale``.
+    It is also possible to write to this property. This has the same behavior as
+    writing ``window-scale``. Note that writingto ``current-window-scale`` will
```
Comment (kasper93) [inline DOCS/man/input.rst:2714]:
```suggestion
    writing ``window-scale``. Note that writing to ``current-window-scale`` will
```

File: DOCS/interface-changes/deprecate-current-window-scale-write.txt (hunk 2 above)
Comment (Dudemanguy) [reply DOCS/interface-changes/deprecate-current-window-scale-write.txt:1]: oops, I forgot to delete this

File: DOCS/man/input.rst [hunk 4]
```diff
@@ -2702,30 +2702,6 @@ Property list
     enabled, or after precise seeking). Files with imprecise timestamps (such
     as Matroska) might lead to unstable results.
 
-``window-scale`` (RW)
```
Comment (kasper93) [inline DOCS/man/input.rst:2705]: is removed from docs intentionally?

**Comment by kasper93** [comment]
I think the idea is ok for what it does. Makes sens to apply it for autofit and geometry and it is not easy to updated those based on current state of window.

`window-minimized` and `window-maximized` are very similar and can be updated externally. They are updated directly to avoid the issue, see:
https://github.com/mpv-player/mpv/blob/6a8b130c6f90f21e5da9bde1da80409976aa2207/video/out/wayland_common.c#L1087-L1089 https://github.com/mpv-player/mpv/blob/6a8b130c6f90f21e5da9bde1da80409976aa2207/video/out/w32_common.c#L1427-L1429

For`window scale we have two props/options `current-window-scale` and `window-scale`. I think this is main issue with this PR, should `window-scale` be refactored and made into one?

File: DOCS/man/input.rst (hunk 4 above)
Comment (Dudemanguy) [reply DOCS/man/input.rst:2705]: Yeah it's mostly redundant/outdated information. Like the other options, it should be in `DOCS/man/options.rst`

**Review by kasper93 — APPROVED** [review]

**Comment by Dudemanguy** [comment]
>I think this is main issue with this PR, should window-scale be refactored and made into one?

Well if you recall, it originally was one property/option but wm4 split them into two. I believe the rationale was that option values should change their value behind the user's back but it's true that some options do this anyway.

**Pull request merged**